civis/parallel.py

```diff
@@ -0,0 +1,719 @@
+                             "or use \"-1\" to request a default "
+                             "of {} jobs.".format(_ALL_JOBS))
+        return n_jobs
+
+    def abort_everything(self, ensure_ready=True):
+        # This method is called when a job has raised an exception.
+        # In that case, we're not going to finish computations, so
+        # we should free up Platform resources in any remaining jobs.
+        self.executor.cancel_all()
+        if not ensure_ready:
+            self.executor.shutdown(wait=False)
+
+    def apply_async(self, func, callback=None):
+        """Schedule func to be run
+        """
+        # Upload the job runner script if needed.
+        if self.run_file_id is None and not self.using_template:
+            runfunc_name = "run_joblib_func.py"
+            runner_local_path = os.path.join(_THIS_DIR, runfunc_name)
+            with open(runner_local_path) as f:
+                self.run_file_id = civis.io.file_to_civis(f, runfunc_name,
+                                                          client=self._client)
+                log.debug("Uploaded job runner script to File %d",
+                          self.run_file_id)
+
+        # Serialize func to a temporary file and upload it to a Civis File.
+        # Make the temporary files expire in a week.
+        expires_at = (datetime.now() + timedelta(days=7)).isoformat()
+        with TemporaryDirectory() as tempdir:
+            temppath = os.path.join(tempdir, "civis_joblib_backend_func")
+            # compress=3 is a compromise between space and read/write times
+            # (https://github.com/joblib/joblib/blob/18f9b4ce95e8788cc0e9b5106fc22573d768c44b/joblib/numpy_pickle.py#L358).
+            joblib.dump(func, temppath, compress=3)
+            with open(temppath, "rb") as tmpfile:
+                func_file_id = \
+                    civis.io.file_to_civis(tmpfile,
+                                           "civis_joblib_backend_func",
+                                           expires_at=expires_at,
+                                           client=self._client)
+                log.debug("uploaded serialized function to File: %d",
+                          func_file_id)
+
+            # Use the Civis CLI client to download the job runner script into
+            # the container, and then run it on the uploaded job.
+            # Only download the runner script if it doesn't already
+            # exist in the destination environment.
+            runner_remote_path = "joblib_remote_worker"
+            cmd = ("{setup_cmd} && "
+                   "if command -v {runner_remote_path} >/dev/null; "
+                   "then {runner_remote_path} {func_file_id}; "
+                   "else pip install joblib=={jl_version} && "
+                   "civis files download {run_file_id} {runner_remote_path} "
+                   "&& python {runner_remote_path} {func_file_id}; fi"
```

Comment:
We should probably add `exec` here like we talked about, in front of both `python` and in between `then {runner_remote_path}`.